flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerStartupTest.java

```diff
@@ -0,0 +1,210 @@
+	}
+
+	/**
+	 * Tests that the TaskManagerRunner startup fails synchronously when the memory configuration is wrong.
+	 */
+	@Test
+	public void testMemoryConfigWrong() {
+		try {
+			Configuration cfg = new Configuration();
+			cfg.setBoolean(TaskManagerOptions.MANAGED_MEMORY_PRE_ALLOCATE, true);
+
+			// something invalid
+			cfg.setString(TaskManagerOptions.MANAGED_MEMORY_SIZE, "-42m");
+			try {
+				TaskManagerRunner.startTaskManager(
+					cfg,
+					ResourceID.generate(),
+					rpcService,
+					mock(HighAvailabilityServices.class),
+					mock(HeartbeatServices.class),
+					NoOpMetricRegistry.INSTANCE,
+					mock(BlobCacheService.class),
+					false,
+					mock(FatalErrorHandler.class));
+
+				fail("Should fail synchronously with an exception");
+			} catch (IllegalConfigurationException e) {
+				// splendid!
+			}
+
+			// something ridiculously high
+			final long memSize = (((long) Integer.MAX_VALUE - 1) *
+				MemorySize.parse(TaskManagerOptions.MEMORY_SEGMENT_SIZE.defaultValue()).getBytes()) >> 20;
+			cfg.setString(TaskManagerOptions.MANAGED_MEMORY_SIZE, memSize + "m");
+			try {
+				TaskManagerRunner.startTaskManager(
+					cfg,
+					ResourceID.generate(),
+					rpcService,
+					mock(HighAvailabilityServices.class),
+					mock(HeartbeatServices.class),
+					NoOpMetricRegistry.INSTANCE,
+					mock(BlobCacheService.class),
+					false,
+					mock(FatalErrorHandler.class));
+
+				fail("Should fail synchronously with an exception");
+			} catch (Exception e) {
+				// splendid!
+				assertTrue(e.getCause() instanceof OutOfMemoryError);
+			}
+		} catch(Exception e) {
+			e.printStackTrace();
+			fail(e.getMessage());
+		}
+	}
+
+	/**
+	 * Tests that the TaskManagerRunner startup fails if the network stack cannot be initialized.
+	 */
+	@Test
+	public void testStartupWhenNetworkStackFailsToInitialize() throws Exception {
+		final ServerSocket blocker = blockServerSocket();
+
+		try {
+			final Configuration cfg = new Configuration();
+			cfg.setInteger(TaskManagerOptions.DATA_PORT, blocker.getLocalPort());
+
+			TaskManagerRunner.startTaskManager(
+				cfg,
+				ResourceID.generate(),
+				rpcService,
+				mock(HighAvailabilityServices.class),
+				mock(HeartbeatServices.class),
+				NoOpMetricRegistry.INSTANCE,
+				mock(BlobCacheService.class),
+				false,
+				mock(FatalErrorHandler.class));
+
+			fail("Should throw IOException when the network stack cannot be initialized.");
+		} catch (IOException e) {
+			// ignored
+		} finally {
+			IOUtils.closeQuietly(blocker);
+		}
+	}
+
+	//-----------------------------------------------------------------------------------------------
+
+	private static RpcService createRpcService() {
+		final RpcService rpcService = mock(RpcService.class);
+		when(rpcService.getAddress()).thenReturn(LOCAL_HOST);
+		return rpcService;
+	}
+
+	private static ServerSocket blockServerSocket() {
+		try {
+			return new ServerSocket(0, 50, InetAddress.getByName(LOCAL_HOST));
+		} catch (IOException e) {
```

Comment:
rethrow the exception instead; if no socket can be allocated it is likely that this code is broken.

Reply:
@zentol In my opinion, if no server could be created with arbitrary ports, which does not mean the Flink code is broken, we should just skip this test.

Reply:
I disagree. If a piece of code is not capable of allocating a random port, which is more likely?
a) All ports in the system are allocated
b) The allocating code is broken in some way
?

Reply:
I think the 'b' answer, "The allocating code is broken in some way", is more likely to happen. However, I don't think this could convince me since we just want a port to be occupied by a running server, and Java built-in `ServerSocket` is a good choice. If an IO error happens when we want to open the socket, this should have no relationship with `TaskManagerRunner` to start up. I don't think the unexpected exception should cause the  test to fail. Current `ZooKeeperLeaderRetrievalTest#testConnectingAddressRetrievalWithDelayedLeaderElection` also consider this unexpected exception case.

Reply:
IMO assumptions should be used very sparingly, as there's the risk of out-right disabling a test completely, without anyone being informed about it. I get your point and agree conceptually, but to this day I have not actually encountered a test failure that would match this case.

And a _lot_ of tests fail if no socket can be allocated, for example those that use a `RestServerEndpoint` or `MiniClusterResource`. In fact, virtually all tests that start a JobManager (or any actor for that matter) will also fail if akka cannot allocate any port.

Reply:
I think allocating sockets in the example you provided actually played a real role. That is to say, the actor created is what we want to test. However, the socket created in `TaskManagerRunnerStartupTest` here was not what we want to test. They're different in this scope.

I think it's not so worthy to spend too much time of discussion on a unit test refractory, especially discarding about what exception to throw out. If you want to stick to that, I think it's also okay for me to change the logic.